File: tests/test_jpeg.py
#!/usr/bin/env python3
"""The JPEG marker walker: geometry and completeness without a decoder.

These matter because both are load-bearing. jpeg_dimensions is how vcap-verify notices
that an HDMI source changed resolution mid-session, and looks_like_jpeg is how a partial
frame gets flagged instead of silently entering a dataset.
"""
from __future__ import annotations

import sys

from fake import make_jpeg

from vcap.frame import jpeg_dimensions, looks_like_jpeg

FAILURES: list[str] = []


def check(name: str, got, want) -> None:
    if got != want:
        FAILURES.append(f"{name}: got {got!r}, want {want!r}")


def main() -> int:
    check("dimensions of a 64x48 frame", jpeg_dimensions(make_jpeg(64, 48)), (64, 48))
    check("dimensions of 1920x1080", jpeg_dimensions(make_jpeg(1920, 1080)),
          (1920, 1080))
    check("dimensions of 2560x1600", jpeg_dimensions(make_jpeg(2560, 1600)),
          (2560, 1600))

    # A frame truncated before the header has no geometry to report, and saying so is the
    # difference between flagging it and recording a wrong size.
    truncated = make_jpeg(1920, 1080)[:4]
    check("truncated before SOF yields None", jpeg_dimensions(truncated), None)
    check("empty input yields None", jpeg_dimensions(b""), None)
    check("not a JPEG yields None", jpeg_dimensions(b"\x89PNG\r\n\x1a\n"), None)

    check("complete JPEG is complete", looks_like_jpeg(make_jpeg(64, 48)), True)
    check("frame without EOI is incomplete",
          looks_like_jpeg(make_jpeg(64, 48, complete=False)), False)
    # The common case by far: the stream-start fragment, which has a perfectly valid EOI
    # and is missing its head. A check that tested only the tail -- the intuitive choice,
    # since signal loss costs the tail -- would pass this and let an undecodable frame
    # into a dataset unflagged.
    headless = make_jpeg(64, 48, headless=True)
    check("a headless fragment is not a whole JPEG", looks_like_jpeg(headless), False)
    check("and it does end with a valid EOI",
          headless[-2:], b"\xff\xd9")
    check("a headless fragment has no readable geometry",
          jpeg_dimensions(headless), None)
    check("empty is not a JPEG", looks_like_jpeg(b""), False)
    check("SOI alone is not a JPEG", looks_like_jpeg(b"\xff\xd8"), False)

    # Padding between segments is legal and encoded as repeated 0xFF. A walker that does
    # not skip it stops at the first pad and reports no geometry.
    padded = bytearray(make_jpeg(800, 600))
    padded[2:2] = b"\xff\xff\xff"
    check("padding bytes before SOF are skipped",
          jpeg_dimensions(bytes(padded)), (800, 600))

    # DHT shares the 0xC0-0xCF range with the start-of-frame markers but is not one. A
    # walker that treats 0xC4 as an SOF reads its table bytes as a resolution.
    with_dht = bytearray(b"\xff\xd8")
    with_dht += b"\xff\xc4\x00\x06\x00\x01\x02\x03"
    with_dht += make_jpeg(320, 240)[2:]
    check("DHT is not mistaken for a start-of-frame",
          jpeg_dimensions(bytes(with_dht)), (320, 240))

    for line in FAILURES:
        print(f"  FAIL  {line}")
    if not FAILURES:
        print("  ok    jpeg marker walking")
    return 1 if FAILURES else 0


if __name__ == "__main__":
    sys.exit(main())

File: vcap_cpp/tests/test_jpeg.cpp
// SPDX-License-Identifier: MIT
//
// The JPEG marker walker. Mirrors tests/test_jpeg.py, including the cases that were got
// wrong there: a fragment missing its head is the common failure on this card, and a
// Huffman table shares the marker range with start-of-frame.
#include <cstdint>
#include <cstdio>
#include <string>
#include <vector>

#include "vcap/frame.hpp"

namespace {

int failures = 0;

void check(const char* name, bool got, bool want) {
    if (got != want) {
        ++failures;
        std::printf("  FAIL  %s: got %s, want %s\n", name, got ? "true" : "false",
                    want ? "true" : "false");
    }
}

void check_dims(const char* name, const std::vector<std::uint8_t>& d,
                bool want_ok, std::uint32_t want_w = 0, std::uint32_t want_h = 0) {
    std::uint32_t w = 0, h = 0;
    const bool ok = vcap::jpeg_dimensions(d.data(), d.size(), &w, &h);
    if (ok != want_ok || (ok && (w != want_w || h != want_h))) {
        ++failures;
        std::printf("  FAIL  %s: got ok=%d %ux%u, want ok=%d %ux%u\n", name, int(ok), w, h,
                    int(want_ok), want_w, want_h);
    }
}

// A structurally valid JPEG: SOI, a baseline SOF0, filler, and optionally EOI. Not
// decodable -- no Huffman table or scan data -- but it has exactly the structure the
// marker walker reads.
std::vector<std::uint8_t> make_jpeg(std::uint32_t w, std::uint32_t h,
                                    bool complete = true, bool headless = false) {
    std::vector<std::uint8_t> d{0xFF, 0xD8};
    const std::uint8_t sof[] = {0xFF, 0xC0, 0x00, 0x11, 0x08,
                                std::uint8_t(h >> 8), std::uint8_t(h & 0xFF),
                                std::uint8_t(w >> 8), std::uint8_t(w & 0xFF),
                                0x03, 0x01, 0x22, 0x00, 0x02, 0x11, 0x01, 0x03, 0x11, 0x01};
    d.insert(d.end(), sof, sof + sizeof sof);
    d.insert(d.end(), {0xFF, 0xFE, 0x00, 0x42});
    d.insert(d.end(), 0x40 - 2, 0x00);
    if (complete) d.insert(d.end(), {0xFF, 0xD9});
    if (headless) d.erase(d.begin(), d.begin() + std::ptrdiff_t(d.size() / 2));
    return d;
}

}  // namespace

int main() {
    check_dims("dimensions of 1920x1080", make_jpeg(1920, 1080), true, 1920, 1080);
    check_dims("dimensions of 2560x1600", make_jpeg(2560, 1600), true, 2560, 1600);
    check_dims("empty input", {}, false);
    check_dims("not a JPEG", {0x89, 'P', 'N', 'G'}, false);

    check("complete JPEG is complete", vcap::looks_like_jpeg(make_jpeg(64, 48)), true);
    check("frame without EOI is incomplete",
          vcap::looks_like_jpeg(make_jpeg(64, 48, false)), false);

    // The common case on this card by far: the stream-start fragment, which has a
    // perfectly valid EOI and is missing its head. A check testing only the tail -- the
    // intuitive choice, since signal loss costs the tail -- would pass this and let an
    // undecodable frame into a dataset unflagged.
    const auto headless = make_jpeg(64, 48, true, true);
    check("a headless fragment is not a whole JPEG", vcap::looks_like_jpeg(headless), false);
    check("and it does end with a valid EOI",
          headless.size() >= 2 && headless[headless.size() - 2] == 0xFF &&
              headless.back() == 0xD9, true);
    check_dims("a headless fragment has no readable geometry", headless, false);

    // DHT shares the 0xC0-0xCF range with the start-of-frame markers but is not one. A
    // walker that treats 0xC4 as an SOF reads its table bytes as a resolution.
    std::vector<std::uint8_t> with_dht{0xFF, 0xD8, 0xFF, 0xC4, 0x00, 0x06, 0x00, 0x01, 0x02, 0x03};
    const auto tail = make_jpeg(320, 240);
    with_dht.insert(with_dht.end(), tail.begin() + 2, tail.end());
    check_dims("DHT is not mistaken for a start-of-frame", with_dht, true, 320, 240);

    if (failures == 0) std::printf("  ok    jpeg marker walking\n");
    return failures ? 1 : 0;
}
